plugins/com.google.cloud.tools.eclipse.usagetracker/src/com/google/cloud/tools/eclipse/usagetracker/AnalyticsPingManager.java

```diff
@@ -0,0 +1,158 @@
+  // https://developers.google.com/analytics/devguides/collection/protocol/v1/reference
+  @SuppressWarnings("serial")
+  private static final Map<String, String> STANDARD_PARAMETERS = Collections.unmodifiableMap(
+      new HashMap<String, String>() {
+        {
+          put("v", "1");  // Google Analytics Measurement Protocol version
+          put("tid", Constants.ANALYTICS_TRACKING_ID);  // tracking ID
+          put("ni", "0");  // Non-interactive? Report as interactive.
+          put("t", "pageview");  // Hit type
+          // "cd??" are for Custom Dimensions in GA.
+          put("cd21", "1");  // Yes, this ping is a virtual "page".
+          put("cd16", "0");  // Internal user? No.
+          put("cd17", "0");  // User signed in? We will ignore this.
+        }
+      });
+
+  private static boolean isTrackingIdDefined() {
+    return Constants.ANALYTICS_TRACKING_ID != null
+        && Constants.ANALYTICS_TRACKING_ID.startsWith("UA-");
+  }
+
+  private static String getAnonymizedClientId() {
+    String clientId = "0";  // For the extremely unlikely event of getNode() failure.
+
+    IEclipsePreferences prefs = ConfigurationScope.INSTANCE.getNode(ANALYTICS_PREFERENCE_QUALIFIER);
+    if (prefs != null) {
+      clientId = prefs.get(PREF_KEY_CLIENT_ID, null);
+      if (clientId == null) {
+        // Use the current time in milliseconds as a client ID and store it back to the preferences.
+        clientId = UUID.randomUUID().toString();
+        prefs.put(PREF_KEY_CLIENT_ID, clientId);
+      }
+    }
+    return clientId;
+  }
+
+  private static boolean hasUserOptedIn() {
+    IEclipsePreferences prefs = ConfigurationScope.INSTANCE.getNode(ANALYTICS_PREFERENCE_QUALIFIER);
+    if (prefs != null) {
+      return prefs.getBoolean(PREF_KEY_USER_CONSENT, false);
+    }
+    return false;
+  }
+
+  public static void sendPing(String eventType, String eventName,
+      String metadataKey, String metadataValue) {
+    Map<String, String> parametersMap = new HashMap<>(STANDARD_PARAMETERS);
+
+    parametersMap.put("cid", getAnonymizedClientId());
+    parametersMap.put("cd19", eventType);
+    parametersMap.put("cd20", eventName);
+
+    String virtualPageUrl = "/virtual/" + APPLICATION_NAME + "/" + eventType + "/" + eventName;
+    parametersMap.put("dp", virtualPageUrl);
+
+    if (metadataKey != null) {
+      // Event metadata are passed as a (virtual) page title.
+      String virtualPageTitle = metadataKey + "=";
+      if (metadataValue != null) {
+        virtualPageTitle += metadataValue;
+      } else {
+        virtualPageTitle += "null";
+      }
+
+      parametersMap.put("dt", virtualPageTitle);
+    }
+
+    sendPostRequest(parametersMap);
+  }
+
+  private static void sendPostRequest(Map<String, String> parametersMap) {
+    if (Platform.inDevelopmentMode() || !isTrackingIdDefined() || !hasUserOptedIn()) {
+      return;
+    }
+
+    String parametersString = getParametersString(parametersMap);
```

Comment:
can you move this into sendPing so a one method adds all the parameters and one actually sends the request?